### ISSUE TITLE
Returning instance of different class after parsing a model

### ISSUE THREAD

**Issue by SaschaSchlemmer** [samuelcolvin/pydantic#1267]
My use case is settings parsing; take as an example 

```python
class URI_List(BaseModel):
    urls: List[AnyHttpUrl]
```
for which I'd like to provide a shortcut for when the base url is always the same

```python
class URL_with_Resources(BaseModel):
    base_url: AnyHttpUrl
    resources: List[str]
```

This is then used in 

```python
class Settings(BaseSettings):
    uris: Union[URI_List, URL_with_Resources]
```

From my application I'd then like to be able and get a list of urls from the settings regardless of how they where provided (in this case via environment variables)

```python
for uri in Settings().uris:
    # do stuff
```

To that end I now (ab-)use `root_validator` to build proper urls from `base_url` and `resources` in `URL_with_Resources` and return an appropriate dictionary. Something like the following (just with proper url processing and additional validation logic perhaps)

```python
class URL_with_Resources(BaseModel):
    base_url: AnyHttpUrl
    resources: List[str]
        
    @root_validator
    def build_urls(cls, values):
        base = values['base_url']
        resources = values['resources']
        return dict(urls=[base+resource for resource in resources])
```

This works but appears very hacky to me and I'd much prefer if there were a way I could return an instance of `URI_List` from `root_validator` or another class method on `URL_with_Resources` instead.

----
Basically the following should return identical objects (maybe instances of a third class `URIs`) so that my application code is not coupled in to the way settings are parsed. 

```python
URI_List(urls=[
    'http://www.example.com/1', 
    'http://www.example.com/2', 
    'http://www.example.com/3']
    )

URL_with_Resoureces(base_url="http://www.example.com", resources=['1', '2', '3'])
```

**Comment by samuelcolvin** [comment]
options:
1. add a method to the settings class (perhaps decorated with `@property`) to build the list of URLs
2. wait for me to properly implement "computed fields" #935 - no timeline on this
3. stick with your `root_validator` approach, I agree it looks hacky but it works
4. you could also use a normal field validator with `pre=True, always=True` to calculate the value for a fields using the `values` argument. This is sort of shown [here](https://pydantic-docs.helpmanual.io/usage/validators/)
